core/src/Cabana_CommunicationPlan.hpp

```diff
@@ -26,6 +28,150 @@
+    Kokkos::View<int*,device_type> _neighbor_counts;
+    Kokkos::View<size_type*,device_type> _neighbor_ids;
+
+    CountSends( ExportRankView element_export_ranks,
+                const int comm_size )
+        : _element_export_ranks( element_export_ranks )
+        , _neighbor_counts( "neighbor_counts", comm_size )
+        , _neighbor_ids(
+            Kokkos::ViewAllocateWithoutInitializing("neighbor_ids"),
+            element_export_ranks.extent(0) )
+    {}
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()( const size_type i ) const
+    {
+        if ( _element_export_ranks(i) >= 0 )
+            _neighbor_ids(i) = Kokkos::atomic_fetch_add(
+                &_neighbor_counts(_element_export_ranks(i)), 1 );
+    }
+
+    void reduce()
+    {}
+};
+#endif // end Cabana_ENABLE_Cuda
+
+// Non-CUDA version. Use duplication.
+template<class ExecutionSpace, class MemorySpace, class ExportRankView>
+struct CountSends
+{
+    Kokkos::Experimental::UniqueToken<
+        ExecutionSpace, Kokkos::Experimental::UniqueTokenScope::Global> _unique_token;
+    ExportRankView _element_export_ranks;
+    using device_type = Kokkos::Device<ExecutionSpace,MemorySpace>;
+    using size_type = typename MemorySpace::size_type;
+    Kokkos::View<int*,device_type> _neighbor_counts;
+    Kokkos::View<size_type*,device_type> _neighbor_ids;
+    Kokkos::View<int**,device_type> _neighbor_counts_dup;
+    Kokkos::View<size_type**,device_type> _neighbor_ids_dup;
+
+    CountSends( ExportRankView element_export_ranks,
+                const int comm_size )
+        : _unique_token()
+        , _element_export_ranks( element_export_ranks )
+        , _neighbor_counts(
+            Kokkos::ViewAllocateWithoutInitializing("neighbor_counts"),
+            comm_size )
+        , _neighbor_ids(
+            Kokkos::ViewAllocateWithoutInitializing("neighbor_ids"),
+            element_export_ranks.extent(0) )
+        , _neighbor_counts_dup( "neighbor_counts",
+                                _unique_token.size(), comm_size )
+        , _neighbor_ids_dup(
+            "neighbor_ids",
+            _unique_token.size(), element_export_ranks.extent(0) )
+    {}
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()( const size_type i ) const
+    {
+        // Get the thread id.
+        auto thread_id = _unique_token.acquire();
+
+        // Do the duplicated fetch-add. If this is a valid element id
+        // increment the send count for this rank. Add the incremented count
+        // as the thread-local neighbor id. This is too big by one (because we
+        // use the prefix increment operator) but we want a non-zero value so
+        // we can later find which thread this element was located on because
+        // we are always guaranteed a non-zero value. We will subtract this
+        // value later.
+        if ( _element_export_ranks(i) >= 0 )
+            _neighbor_ids_dup(thread_id,i) =
+                ++_neighbor_counts_dup( thread_id, _element_export_ranks(i) );
```

Comment:
Because I want them to be zero if I don't find anything for the total count sum later on in `reduce`. If I initialized to `-1` I'd have to do the addition there instead.